### PR TITLE
Refactors AI / camera eyes and slows holopad holograms to walk speed

### PR DESCRIPTION
**Pull request by asciodev**
## What Does This PR Do
This change completes a refactor of AI eyes, which were previously used by xenobio consoles, syndicate and abductor camera consoles, shuttle docking computers, holograms, and, of course, the AI. Duplicated logic has been extracted to an abstract base mob, /mob/camera/eye, from which new types for each of the above now derive.  Functionality is largely the same, with only a few minor cosmetic differences (i.e. camera eyes are now appropriately named given their type and user), as well as a quality-of-life enhancement for holograms, slowing their movement speed to base run speed to prevent users from accidentally zooming out of calls. A toggle for switching between this behavior and full speed has also been added to AI Commands, called "Toggle Fast Holograms", disabled by default.

## Why It's Good For The Game
* Duplicated logic sporadically spread across files is hard to follow; maintainability is good.
* Holograms are currently nearly unusable by crew due to how easy it is to zoom out of holopad range at AI eye speeds. This change fixes that.

## Testing
* Using a test server, I logged in as a Head of Personnel, VV'd the nearest holopad, and set `force_answer_call` to 1. I then called the Bridge, and upon being auto-answered, I was given sight and control over a hologram on the bridge, as expected. It was no longer zoomy. I left the holopad range and was disconnected from the call, regaining sight and control over my character. I was also able to exit a call using the associated action button.
* I then made myself an AI, and verified that I could still move around the station, switch various cameras, view my core, turn on and off camera lights, turn on and off acceleration, and set and use saved camera locations. I activated a holopad and moved around, which worked as expected, and identically in function to before this PR, except no longer zoomy. Upon leaving range of nearby holopads, my hologram disappeared, as expected, and my speed returned to normal AI eye speed.
* I wiped my core and sent myself back to lobby, and joined as a Solar Federation General. I walked to the specops navigation computer on the port-side shuttle and verified that I could still create a custom shuttle destination, and that the turf overlays designating valid/invalid destination areas still showed up, and that I could not designate invalid transit locations.
* I spawned an advanced camera console (the kind that appears in the Syndicate Research Base) and used it. I was able to jump to various cameras and move around.
* I made myself an abductor scientist and teleported to the abductor ship. I used the abductor camera console, and was able to jump to various cameras, move the camera around, set teleport targets, teleport myself in, teleport Pun Pun in, and my vision was blocked by static where there were no active cameras.
* I teleported myself to Xenobiology, and used one of their consoles. I was able to pick up slimes, place monkeys, and exit the camera via action button. The camera was still confined to the allowed xenobiology area, and attempting to jump to areas that were not in the allowed area still bounced back to xenobiology.

## Changelog
:cl: Ascio
add: Camera eyes not belonging to the AI now have descriptive names, allowing easier orbiting
add: Added a "Fast Holograms" toggle to AI commands, which sets holopad holograms to camera eye speed (disabled by default)
tweak: Slowed holopad holograms to base run speed by default to prevent accidental holocall hangups
/:cl:

### CODE REVIEW

File: code/__DEFINES/mob_defines.dm [hunk 1]
```diff
@@ -266,7 +266,8 @@
 #define isAutoAnnouncer(A)	(istype((A), /mob/living/automatedannouncer))
 
 #define iscameramob(A)	(istype((A), /mob/camera))
-#define isAIEye(A)		(istype((A), /mob/camera/aiEye))
+#define isAIEye(A)		(istype((A), /mob/camera/eye))
+#define isholo(A)			(istype((A), /mob/camera/eye/hologram))
```
Comment (GDNgit) [inline code/__DEFINES/mob_defines.dm:270]:
```suggestion
#define isholo(A)			(istype((A), /mob/camera/eye/hologram))
```
probably best to rename this to ishologram or something due to holoparas also using the holo word

File: code/game/machinery/hologram.dm [hunk 2]
```diff
@@ -335,8 +336,8 @@ GLOBAL_LIST_EMPTY(holopads)
 
 //Can we display holos there
 //Area check instead of line of sight check because this is a called a lot if AI wants to move around.
-/obj/machinery/hologram/holopad/proc/validate_location(turf/T,check_los = FALSE)
-	if(T.z == z && get_dist(T, src) <= holo_range && T.loc == get_area(src))
+/obj/machinery/hologram/holopad/proc/validate_location(turf/T, extended_range = 0)
+	if(T.z == z && get_dist(T, src) <= holo_range + extended_range && T.loc == get_area(src))
```
Comment (GDNgit) [inline code/game/machinery/hologram.dm:340]:
```suggestion
	if(T.z == z && (get_dist(T, src) <= holo_range + extended_range) && T.loc == get_area(src))
```
massively loaded function so it's best to separate via parentheses

File: code/modules/mob/camera/eye.dm [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	var/max_sprint = 50
+	var/sprint_threshold = 20
+	var/acceleration_rate = 0.5
+	var/sprint = 10
+	var/cooldown = 0
+	var/cooldown_rate = 5
+	var/acceleration = 1
+
+/mob/camera/eye/Initialize(mapload, name, origin, mob/living/user)
+	..()
```
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:35]:
```suggestion
	. = ..()
```
Important to return an init hint

File: code/modules/mob/camera/eye.dm [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+/mob/camera/eye/setLoc(T, hide_unseen_with_static = 1)
+	if(user)
+		T = get_turf(T)
+		..(T)
+		update_visibility()
+		refresh_visible_icon()
+
+/mob/camera/eye/Move()
+	return 0
```
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:76]:
```suggestion
	return FALSE
```

File: code/modules/mob/camera/eye.dm [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		var/mob/living/silicon/ai/AI = usr
+		if(AI.eyeobj && (AI.client.eye == AI.eyeobj) && (AI.eyeobj.z == z))
+			AI.cameraFollow = null
+			if(isturf(loc) || isturf(src))
+				AI.eyeobj.setLoc(src)
+
+/mob/camera/eye/proc/GetViewerClient()
+	if(user)
+		return user.client
+	return null
```
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:89]:
```suggestion
	return user?.client
```
maybe good to just nuke this proc and call it directly

File: code/modules/mob/camera/eye.dm [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+/mob/camera/eye/proc/RemoveImages()
+	var/client/C = GetViewerClient()
+	if(C)
+		for(var/V in visibleCameraChunks)
+			var/datum/camerachunk/chunk = V
+			C.images -= chunk.obscured
+		if(visible_icon)
+			C.images -= user_image
```
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:98]:
```suggestion
	var/client/C = GetViewerClient()
	if(!C)
		return
	for(var/datum/camerachunk/chunk as anything in visibleCameraChunks)
		C.images -= chunk.obscured
	if(visible_icon)
		C.images -= user_image
```

File: code/modules/mob/camera/eye.dm [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		user.reset_perspective(user_previous_remote_control)
+		user.remote_control = user_previous_remote_control
+		user_previous_remote_control = null
+	user = null
+
+/mob/camera/eye/proc/give_control(mob/user)
+	if(!istype(user))
+		return
+	release_control()
+	src.user = user
```
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:117]: source stuff as well here

File: code/modules/mob/camera/eye.dm [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+	user.remote_control = src
+	user.reset_perspective(src)
+
+/mob/camera/eye/proc/rename_camera(new_name)
+	name = "Camera Eye ([new_name])"
+
+/mob/camera/eye/proc/release_chunks()
+	for(var/V in visibleCameraChunks)
+		var/datum/camerachunk/chunk = V
+		chunk.remove(src)
```
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:130]:
```suggestion
	for(var/datum/camerachunk/chunk as anything in visibleCameraChunks)
		chunk.remove(src)
```

File: code/modules/mob/camera/eye.dm [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+	cooldown = world.timeofday + cooldown_rate
+	if(acceleration)
+		sprint = min(sprint + acceleration_rate, max_sprint)
+	else
+		sprint = initial
+
+/mob/camera/eye/hear_say(list/message_pieces, verb = "says", italics = 0, mob/speaker = null, sound/speech_sound, sound_vol, sound_frequency, use_voice = TRUE)
+	if(relay_speech)
+		var/mob/M = user
+		M.hear_say(message_pieces, verb, italics, speaker, speech_sound, sound_vol, sound_frequency)
```
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:157]:
```suggestion
		user.hear_say(message_pieces, verb, italics, speaker, speech_sound, sound_vol, sound_frequency)
```

File: code/modules/mob/camera/eye/hologram_eye.dm [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+		if(loc == step)
+			delay = diag_delay
+	user.last_movement = world.time
+	delay = TICKS2DS(-round(-(DS2TICKS(delay))))
+	user.client.move_delay += delay
+
+/mob/camera/eye/hologram/validate_active_cameranet()
+	..(TRUE)
+
+/mob/camera/eye/hologram/setLoc(T, delay, initial = 0)
```
Comment (GDNgit) [inline code/modules/mob/camera/eye/hologram_eye.dm:41]: lord this proc

File: code/modules/mob/camera/eye.dm [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+/mob/camera/eye/relaymove(mob/user,direct)
+	var/initial = initial(sprint)
+
+	if(cooldown && cooldown < world.timeofday)
+		sprint = initial
+
+	for(var/i = 0; i < max(sprint, initial); i += sprint_threshold)
+		var/turf/step = get_turf(get_step(src, direct))
+		if(step)
+			src.setLoc(step)
```
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:146]:
```suggestion
			setLoc(step)
```

File: code/modules/mob/camera/eye.dm [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+	release_chunks()
+	return ..()
+
+/mob/camera/eye/relaymove(mob/user,direct)
+	var/initial = initial(sprint)
+
+	if(cooldown && cooldown < world.timeofday)
+		sprint = initial
+
+	for(var/i = 0; i < max(sprint, initial); i += sprint_threshold)
```
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:143]:
```suggestion
	for(var/i = 0; i < max(sprint, initial); i += sprint_threshold)
```
you can probably just replace this with a to loop

File: code/modules/mob/camera/eye.dm (hunk 11 above)
Comment (GDNgit) [inline code/modules/mob/camera/eye.dm:146]:
```suggestion
		var/turf/next_step= get_turf(get_step(src, direct))
		if(next_step)
			src.setLoc(next_step)
```
don't use reserved keywords on vars

File: code/modules/mob/camera/eye/ai_eye.dm [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+	name = "Inactive AI Eye"
+	icon = 'icons/mob/ai.dmi'
+	icon_state = "eye"
+	var/mob/living/silicon/ai/ai = null
+
+/mob/camera/eye/ai/Initialize(mapload, name, origin, mob/living/user)
+	..()
+	ai = user
+	if(isAIEye(ai.eyeobj))
+		to_chat(user, "<span class='warning'>ERROR: There is an AI Eye already assigned.</span>")
```
Comment (GDNgit) [inline code/modules/mob/camera/eye/ai_eye.dm:11]:
```suggestion
		to_chat(user, "<span class='warning'>ERROR: There is an AI Eye already assigned.</span>")
```
stack trace this

File: code/modules/mob/camera/eye/shuttle_navigator.dm [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+/mob/camera/eye/shuttle_docker/Initialize(mapload, name, origin, mob/living/user)
+	..()
+	setLoc(first_active_camera())
+
+/mob/camera/eye/shuttle_docker/rename_camera(new_name)
+	name = "Shuttle Docker Camera Eye ([new_name])"
+
+/mob/camera/eye/shuttle_docker/update_visibility()
+	return 0
```
Comment (GDNgit) [inline code/modules/mob/camera/eye/shuttle_navigator.dm:16]:
```suggestion
	return FALSE
```

File: code/modules/mob/camera/eye/shuttle_navigator.dm [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+/mob/camera/eye/shuttle_docker/setLoc(T)
+	if(isspaceturf(get_turf(T)) || istype(get_area(T), /area/space) || istype(get_area(T), /area/shuttle))
+		..()
+		var/obj/machinery/computer/camera_advanced/shuttle_docker/console = origin
+		console.checkLandingSpot()
+		return
+	else
+		return
```
Comment (GDNgit) [inline code/modules/mob/camera/eye/shuttle_navigator.dm:25]:
```suggestion
```

File: code/modules/mob/camera/eye/syndicate.dm [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+/mob/camera/eye/syndicate/Initialize(mapload, name, origin, user)
+	..()
+	setLoc(first_active_camera())
+
+/mob/camera/eye/syndicate/rename_camera(new_name)
+	name = "Syndicate Researcher Camera Eye ([new_name])"
+	
+/mob/camera/eye/syndicate/update_visibility()
+	return 0
```
Comment (GDNgit) [inline code/modules/mob/camera/eye/syndicate.dm:13]:
```suggestion
	return FALSE
```

File: code/modules/mob/living/silicon/ai/ai_mob.dm [hunk 17]
```diff
@@ -1477,8 +1476,36 @@ GLOBAL_LIST_INIT(ai_verbs_default, list(
...
+		to_chat(src, "ERROR: Eyeobj not found. Creating new eye...")
+		eyeobj = new /mob/camera/eye/ai(loc, name, src, src)
+
+	eyeobj.setLoc(loc)
+
+/mob/living/silicon/ai/proc/toggle_acceleration()
+	set category = "AI Commands"
+	set name = "Toggle Camera Acceleration"
+
+	if(usr.stat == 2)
```
Comment (GDNgit) [inline code/modules/mob/living/silicon/ai/ai_mob.dm:1504]:
```suggestion
	if(usr.stat == DEAD)
```

File: code/modules/mob/living/silicon/ai/ai_mob.dm [hunk 18]
```diff
@@ -1477,8 +1476,36 @@ GLOBAL_LIST_INIT(ai_verbs_default, list(
...
+	view_core()
+
+/mob/living/silicon/ai/proc/view_core()
+
+	current = null
+	cameraFollow = null
+	unset_machine()
+
+	if(src.eyeobj && src.loc)
+		src.eyeobj.loc = src.loc
```
Comment (GDNgit) [inline code/modules/mob/living/silicon/ai/ai_mob.dm:1493]:
```suggestion
	if(eyeobj && loc)
		eyeobj.loc = loc
```

File: code/modules/mob/mob_movement.dm [hunk 19]
```diff
@@ -93,12 +93,17 @@
...
+		var/mob/living/silicon/ai/ai = mob
+		var/mob/camera/eye/ai/eye = ai.eyeobj
+		if(istype(eye) && !istype(ai.remote_control))
+			ai.remote_control = eye
+			return eye.relaymove(mob, direct)
+		if(!istype(eye) && !istype(mob.loc, /obj/item/aicard))
+			eye = new /mob/camera/eye/ai(mob.loc, ai.name, ai, ai)
+			if(istype(eye))
+				return eye.relaymove(mob, direct)
+		return 0 // If the AI is outside of its eye or a mech (e.g. carded), it can't move
```
Comment (GDNgit) [inline code/modules/mob/mob_movement.dm:106]:
```suggestion
		return FALSE // If the AI is outside of its eye or a mech (e.g. carded), it can't move
```

**Review by GDNgit — CHANGES_REQUESTED** [review]

File: code/modules/mob/camera/eye/hologram_eye.dm (hunk 10 above)
Comment (asciodev) [reply code/modules/mob/camera/eye/hologram_eye.dm:41]: Is this in reference to the `validate_active_cameranet` overloading, or the ripped `relaymove` code?

If the former, I understand - it made sense to me while in the weeds of this to try to get rid of as many variable-based checks as possible, due to their inconsistent use. I felt that using procs and overloading them down the hierarchy was less fragile. I can understand how it's unclear at a glance though, and am happy to change it back to var-based checks.

If the latter, my defense is that "it works", and attempting to DRY it up would involve touching `mob_movement.dm` code that doesn't directly concern the AI, which felt a bit like scope creep for something that appears in the code only twice and in very different contexts.

**Comment by Daylight2** [comment]
I personally liked being able to zoom away as AI if you were urgently needed as whilst in a holopad, although I suppose that run speed isn't *that* slow. AI players are probably used to the high speed, however, and the fix to messing up is just triple clicking for them, instead of requesting a whole new call.

**Comment by asciodev** [comment]
After some discussion on discord I agree that the change could be frustrating for many AI players who are used to the zoominess and have mastered it. Although the AI Commands tab is already rather bloated, I will add a toggle for fast holograms.

File: code/modules/mob/camera/eye.dm [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+	var/acceleration_rate = 0.5
+	var/sprint = 10
+	var/cooldown = 0
+	var/cooldown_rate = 5
+	var/acceleration = 1
+
+/mob/camera/eye/Initialize(mapload, name, origin, mob/living/user)
+	..()
+	src.name = "Camera Eye ([name])"
+	src.origin = origin
```
[the remark this answers is not in the document]
Comment (asciodev) [reply code/modules/mob/camera/eye.dm]: I've changed var names in initialization across eye subtypes. Are the remaining uses of src (e.g. in places like `user.reset_perspective(src)`) acceptable?

File: code/modules/mob/camera/eye/hologram_eye.dm (hunk 10 above)
Comment (asciodev) [reply code/modules/mob/camera/eye/hologram_eye.dm]: I've looked at this with fresh eyes and have refactored it into something more sane.

**Review by FunnyMan3595 — CHANGES_REQUESTED** [review]
Update the PR description to reflect the added setting (and that existing behavior is the default). LGTM otherwise.

**Comment by asciodev** [comment]
There is actually a potential malf interaction that I forgot to check and need to look into before this can be merged. I'm on a new VM and need to redo my dev environment, so it'll be a couple days before I can get to it.

File: code/modules/mob/camera/eye.dm [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+	// Every move otherwise increases sprint by acceleration_rate, until sprint
+	// exceeds sprint_threshold, and the movement rate increases by one per move.
+	// The movement rate is 1 + round(sprint / sprint_threshold).
+	var/max_sprint = 50
+	var/sprint_threshold = 20
+	var/acceleration_rate = 0.5
+	var/sprint = 10
+	var/cooldown = 0
+	var/cooldown_rate = 5
+	var/acceleration = 1
```
Comment (Spaghetti-bit) [inline code/modules/mob/camera/eye.dm:32]: Please autodoc all of these in the future. Functionality seems fine but these need to be autodoc'd
Example: `/// what this var does`

**Review by Spaghetti-bit — COMMENTED** [review]
Few minor changes but functionality seems fine. LGTM

**Comment by asciodev** [comment]
> There is actually a potential malf interaction that I forgot to check and need to look into before this can be merged. I'm on a new VM and need to redo my dev environment, so it'll be a couple days before I can get to it.

Sorry it took me so long to get around to this, I wasn't able to successfully configure a dev environment on the new VM until recently. I verified that this malf interaction was working fine, but discovered that the syndicate advanced console was erroneously giving free xray vision, so I went ahead and reverted that.

I've tested again after merging with master, and everything appears in order.

File: code/game/machinery/hologram.dm [hunk 22]
```diff
@@ -92,6 +92,7 @@ GLOBAL_LIST_EMPTY(holopads)
 	var/ringing = FALSE
 	/// Whether or not the user is currently selecting where to send their call.
 	var/dialling_input = FALSE
+	var/mob/camera/eye/hologram/eye = null
```
Comment (Drsmail) [inline code/game/machinery/hologram.dm:95]:
```suggestion
	var/mob/camera/eye/hologram/eye
```

File: code/modules/mob/camera/eye.dm [hunk 23]
```diff
@@ -0,0 +1,158 @@
+/mob/camera/eye
+	name = "Inactive Camera Eye"
+	icon = 'icons/obj/abductor.dmi'
+	icon_state = "camera_target"
+	alpha = 127
+	invisibility = SEE_INVISIBLE_OBSERVER
+
+	var/list/visibleCameraChunks = list()
+	var/mob/living/user = null
```
Comment (Drsmail) [inline code/modules/mob/camera/eye.dm:9]:
```suggestion
	var/mob/living/user
```

File: code/modules/mob/camera/eye.dm [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+	var/user_previous_remote_control = null
```
Comment (Drsmail) [inline code/modules/mob/camera/eye.dm:10]:
```suggestion
	var/user_previous_remote_control
```

File: code/modules/mob/camera/eye.dm [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+	var/origin = null
```
Comment (Drsmail) [inline code/modules/mob/camera/eye.dm:11]:
```suggestion
	var/origin
```

File: code/modules/mob/camera/eye.dm [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+	var/relay_speech = FALSE
+	var/static_visibility_range = 16
+	// Decides if it is shown by AI Detector or not
+	var/ai_detector_visible = TRUE
+	var/visible_icon = FALSE
+	var/list/networks = list("SS13")
+	var/image/user_image = null
```
Comment (Drsmail) [inline code/modules/mob/camera/eye.dm:18]:
```suggestion
	var/image/user_image
```

File: code/modules/mob/camera/eye/hologram_eye.dm [hunk 27]
```diff
@@ -0,0 +1,49 @@
+/mob/camera/eye/hologram
+	name = "Inactive Hologram Eye"
+	ai_detector_visible = FALSE
+	acceleration = FALSE
+	relay_speech = TRUE
+	var/obj/machinery/hologram/holopad/holopad = null
```
Comment (Drsmail) [inline code/modules/mob/camera/eye/hologram_eye.dm:6]:
```suggestion
	var/obj/machinery/hologram/holopad/holopad
```

File: code/modules/mob/camera/eye/xenobio_eye.dm [hunk 28]
```diff
@@ -0,0 +1,21 @@
+/mob/camera/eye/xenobio
+	name = "Xenobiology Console Camera Eye"
+	visible_icon = TRUE
+	ai_detector_visible = FALSE
+	var/allowed_area = null
```
Comment (Drsmail) [inline code/modules/mob/camera/eye/xenobio_eye.dm:5]:
```suggestion
	var/allowed_area
```

File: code/modules/mob/living/silicon/ai/ai_mob.dm [hunk 29]
```diff
@@ -104,10 +105,8 @@ GLOBAL_LIST_INIT(ai_verbs_default, list(
 	var/obj/machinery/doomsday_device/doomsday_device
 
 	var/obj/machinery/hologram/holopad/holo = null
-	var/mob/camera/ai_eye/eyeobj
-	var/sprint = 10
-	var/cooldown = 0
-	var/acceleration = 1
+	var/mob/camera/eye/ai/eyeobj = null
```
Comment (Drsmail) [inline code/modules/mob/living/silicon/ai/ai_mob.dm:108]:
```suggestion
	var/mob/camera/eye/ai/eyeobj
```

**Review by Drsmail — CHANGES_REQUESTED** [review]
Really like what you did here, didn't have time to chek logick, will give it a second look later.

**Review by FunnyMan3595 — APPROVED** [review]
New description/changelog LGTM.

File: code/datums/holocall.dm [hunk 30]
```diff
@@ -1,18 +1,10 @@
...
-//this datum manages it's own references
-
 /datum/holocall
 	var/mob/living/user	//the one that called
 	var/obj/machinery/hologram/holopad/calling_holopad	//the one that sent the call
 	var/obj/machinery/hologram/holopad/connected_holopad	//the one that answered the call (may be null)
 	var/list/dialed_holopads	//all things called, will be cleared out to just connected_holopad once answered
 
-	var/mob/camera/ai_eye/remote/holo/eye	//user's eye, once connected
+	var/mob/camera/eye/eye //eye, once connected
```
Comment (lewcc) [inline code/datums/holocall.dm:7]:
```suggestion
	///eye, once connected
	var/mob/camera/eye/eye 
```

File: code/modules/mob/camera/eye.dm [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+
+/atom/proc/move_camera_by_click()
+	if(isAI(usr))
+		var/mob/living/silicon/ai/AI = usr
+		if(AI.eyeobj && (AI.client.eye == AI.eyeobj) && (AI.eyeobj.z == z))
+			AI.cameraFollow = null
+			if(isturf(loc) || isturf(src))
+				AI.eyeobj.setLoc(src)
+
+/mob/camera/eye/proc/GetViewerClient()
```
Comment (lewcc) [inline code/modules/mob/camera/eye.dm:91]: Would you be able to change the names of these procs to `camel_case` instead?

File: code/modules/mob/camera/eye.dm [hunk 32]
```diff
@@ -0,0 +1,158 @@
+/mob/camera/eye
+	name = "Inactive Camera Eye"
+	icon = 'icons/obj/abductor.dmi'
+	icon_state = "camera_target"
+	alpha = 127
+	invisibility = SEE_INVISIBLE_OBSERVER
+
+	var/list/visibleCameraChunks = list()
```
Comment (lewcc) [inline code/modules/mob/camera/eye.dm:8]: Could you rename this to `visible_camera_chunks`?

File: code/modules/mob/camera/eye/ai_eye.dm [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+	icon_state = "eye"
+	var/mob/living/silicon/ai/ai = null
+
+/mob/camera/eye/ai/Initialize(mapload, owner_name, camera_origin, mob/living/user)
+	. = ..()
+	ai = user
+	if(isAIEye(ai.eyeobj))
+		stack_trace("Tried to create an AI Eye for [user], but there is one already assigned")
+		qdel(src)
+		return
```
Comment (lewcc) [inline code/modules/mob/camera/eye/ai_eye.dm:13]:
```suggestion
		return INITIALIZE_HINT_QDEL
```

**Comment by asciodev** [comment]
I mass-replaced everything that I could find related to this system from `PascalCase` to `camel_case`, except for TGUI data prop names. I also added autodoc comments to just about every variable and function related to camera eyes. I tested as per the PR description and no unexpected behavior or errors cropped up.

**Review by Burzah — APPROVED** [review]

**Review by FunnyMan3595 — APPROVED** [review]

**Comment by TravisAngeI** [comment]
Something like, mildly annoying but super small that I doubt can be fixed, is that if the AI moves in their camera-view, and I'm pretty sure it's as a result of this PR, and they let go of the key while the game is still trying to move them to that tile, it does like a weird, slow movement, it's mostly a visual thing.

Just figured I would point it out (this is kinda nitpicky), it has no real harm but I was maybe hoping to see if there was possibility of tweaking it to make it not do that? Not sure how code worky.

It does this to holopads aswell, much more noticeably (see second video if they both attach properly)

https://github.com/user-attachments/assets/68c3f16c-5eea-4268-b6aa-4869c1904771

https://github.com/user-attachments/assets/3641d973-b866-4c8b-aaa7-571cfebf92e4

**Comment by TravisAngeI** [comment]
> Something like, mildly annoying but super small that I doubt can be fixed, is that if the AI moves in their camera-view, and I'm pretty sure it's as a result of this PR, and they let go of the key while the game is still trying to move them to that tile, it does like a weird, slow movement, it's mostly a visual thing.

Okay, after testing this out the next shift, this seems to not be a problem anymore? Super weird.
